kernel/ixgbe/src/lib.rs

```diff
@@ -0,0 +1,1192 @@
+        let nic_tx_regs_mapped_page = allocate_memory(mem_base + GENERAL_REGISTERS_1_SIZE_BYTES + RX_REGISTERS_SIZE_BYTES + GENERAL_REGISTERS_2_SIZE_BYTES, TX_REGISTERS_SIZE_BYTES)?;
+        let nic_mac_regs_mapped_page = allocate_memory(mem_base + GENERAL_REGISTERS_1_SIZE_BYTES + RX_REGISTERS_SIZE_BYTES + GENERAL_REGISTERS_2_SIZE_BYTES + TX_REGISTERS_SIZE_BYTES, MAC_REGISTERS_SIZE_BYTES)?;
+        let nic_rx_regs2_mapped_page = allocate_memory(mem_base + GENERAL_REGISTERS_1_SIZE_BYTES + RX_REGISTERS_SIZE_BYTES + GENERAL_REGISTERS_2_SIZE_BYTES + TX_REGISTERS_SIZE_BYTES + MAC_REGISTERS_SIZE_BYTES, RX_REGISTERS_SIZE_BYTES)?;        
+        let nic_regs3_mapped_page = allocate_memory(mem_base + GENERAL_REGISTERS_1_SIZE_BYTES + RX_REGISTERS_SIZE_BYTES + GENERAL_REGISTERS_2_SIZE_BYTES + TX_REGISTERS_SIZE_BYTES + MAC_REGISTERS_SIZE_BYTES + RX_REGISTERS_SIZE_BYTES, GENERAL_REGISTERS_3_SIZE_BYTES)?;
+
+        // Map the memory as the register struct and tie the lifetime of the struct with its backing mapped pages
+        let regs1 = BoxRefMut::new(Box::new(nic_regs1_mapped_page)).try_map_mut(|mp| mp.as_type_mut::<IntelIxgbeRegisters1>(0))?;
+        let regs2 = BoxRefMut::new(Box::new(nic_regs2_mapped_page)).try_map_mut(|mp| mp.as_type_mut::<IntelIxgbeRegisters2>(0))?;
+        let regs3 = BoxRefMut::new(Box::new(nic_regs3_mapped_page)).try_map_mut(|mp| mp.as_type_mut::<IntelIxgbeRegisters3>(0))?;
+        let mac_regs = BoxRefMut::new(Box::new(nic_mac_regs_mapped_page)).try_map_mut(|mp| mp.as_type_mut::<IntelIxgbeMacRegisters>(0))?;
+        
+        // Divide the pages of the Rx queue registers into multiple 64B regions
+        let mut regs_rx = Self::mapped_regs_from_rx_memory(nic_rx_regs1_mapped_page);
+        regs_rx.append(&mut Self::mapped_regs_from_rx_memory(nic_rx_regs2_mapped_page));
+        
+        // Divide the pages of the Tx queue registers into multiple 64B regions
+        let regs_tx = Self::mapped_regs_from_tx_memory(nic_tx_regs_mapped_page);
+            
+        Ok((regs1, regs2, regs3, mac_regs, regs_rx, regs_tx))
+    }
+
+    /// Split the pages where rx queue registers are mapped into multiple smaller memory regions.
+    /// One region contains all the registers for a single queue.
+    fn mapped_regs_from_rx_memory(mp: MappedPages) -> Vec<IxgbeRxQueueRegisters> {
+        const QUEUES_IN_MP: usize = 64;
+        const RX_QUEUE_REGISTERS_SIZE_BYTES: usize = core::mem::size_of::<RegistersRx>();
+        
+        assert!(mp.size_in_bytes() >= QUEUES_IN_MP * RX_QUEUE_REGISTERS_SIZE_BYTES);
+
+        let starting_address = mp.start_address();
+
+        // We share the backing mapped pages among all the queue registers
+        let shared_mp = Arc::new(mp);
+        let mut pointers_to_queues = Vec::with_capacity(QUEUES_IN_MP);
+
+        for i in 0..QUEUES_IN_MP {
+            // This is safe because we have checked that the number of queues we want to partition from these mapped pages fit into the allocated memory,
```

Comment:
We ensure that no memory overlaps. This is the core of the unsafe code and the point where the developer has to ensure correctness.